### PR TITLE
Document Supertraits and `where` clauses

### PR DESCRIPTION
**Pull request by ControlCplusControlV**
As always if my language doesn't match docs style or code examples aren't in depth enough lmk. The mention of the core library Ord isn't implemented yet but planning on getting to it, the examples don't focus on implementing other things but showing how you would, so if they need to be expanded into more real world examples I can do that

### CODE REVIEW

File: docs/src/advanced/generic_types.md [hunk 1]
```diff
@@ -30,6 +30,17 @@ purely shorthand for the sake of ergonomics.
 
 ## Trait Constraints
 
+Important background to know before diving into trait constraints is that the `where` clause can be used to specify needed traits of the generic argument. So when writing something like a Hashmap you may
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:33]:
```suggestion
Important background to know before diving into trait constraints is that the `where` clause can be used to specify the required traits for the generic argument. So, when writing something like a `HashMap` you may
```

File: docs/src/advanced/generic_types.md [hunk 2]
```diff
@@ -30,6 +30,17 @@ purely shorthand for the sake of ergonomics.
...
+want to specify that the generic argument implements a `Hash` trait.
+
+```sway
+fn get_hashmap_key<T>(Key : T) -> b256
+  where T: Hash
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:38]:
```suggestion
    where T: Hash
```

File: docs/src/advanced/generic_types.md [hunk 3]
```diff
@@ -30,6 +30,17 @@ purely shorthand for the sake of ergonomics.
...
+{
+  // Code within here can then call functions assosiated with the Hash trait on Key
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:40]:
```suggestion
    // Code within here can then call functions associated with the Hash trait on Key
```

File: docs/src/advanced/generic_types.md [hunk 4]
```diff
@@ -30,6 +30,17 @@ purely shorthand for the sake of ergonomics.
 
 ## Trait Constraints
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:31]: We should a note in italic at the end of this section that says that
"`where` clauses are still work-in-progress".

**Review by mohammadfawaz — CHANGES_REQUESTED** [review]
Thanks for doing this! When you make the requested changes, I will review again to make sure it's all coherent :)

**Comment by ControlCplusControlV** [comment]
Changes are added, really need to added code spell check to my nvim config

File: docs/src/advanced/generic_types.md [hunk 5]
```diff
@@ -30,6 +30,17 @@ purely shorthand for the sake of ergonomics.
...
 
+Important background to know before diving into trait constraints is that the `where` clause can be used to specify the required traits for the generic argument. So, when writing something like a `HashMap` you may
+want to specify that the generic argument implements a `Hash` trait.
+
+```sway
+fn get_hashmap_key<T>(Key : T) -> b256
+    where T: Hash
+{
+    // Code within here can then call functions associated with the Hash trait on Key
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:40]:
```suggestion
    // Code within here can then call methods associated with the Hash trait on Key
```

File: docs/src/advanced/generic_types.md [hunk 6]
```diff
@@ -108,6 +119,8 @@ struct Foo<T>
 }
 ```
 
+*`where` clauses are still work-in-progress, so some above statements may not be fully implemented*
```
Comment (mohammadfawaz) [inline docs/src/advanced/generic_types.md:122]: Oh I think should actually go at the end of the "Traits Constraints" section instead.

**Review by mohammadfawaz — CHANGES_REQUESTED** [review]
Two more minor things and you're good to go!

**Comment by ControlCplusControlV** [comment]
Changes committed  (Assuming I moved the warning to the right spot)

Should note this closes both #753 and #854

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**